internal/terraform/node_module_variable.go

```diff
@@ -253,7 +253,23 @@ func (n *nodeModuleVariable) evalModuleCallArgument(ctx EvalContext, validateOnl
 		val = cty.UnknownVal(n.Config.Type)
 	}
 
+	// If there is no default, we have to ensure that a null value is allowed
+	// for this variable.
+	if n.Config.Default == cty.NilVal && !n.Config.Nullable && val.IsNull() {
+		// The value cannot be null, and there is no configured default.
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  `Invalid variable value`,
+			Detail:   fmt.Sprintf(`The variable %q is required, but the given value resolved to null`, n.Addr),
```

Comment:
One final nitpick :grimacing: : detail messages should end with a period for consistency with our usual style.

More subjectively, I think we don't normally talk about values being "resolved" so maybe we can just say "but the given value _is_ null".

Reply:
The resolved term was taken from your other variable evaluation diagnostic 😉
```
The resolved value of variable %q is not appropriate:
```

But I agree a more direct "the value is" ends up sounding better here.